### ISSUE TITLE
Does the API expose which request arguments are optional?

### ISSUE THREAD

**Issue by lvh** [aws/aws-sdk-java#965]
Many API calls have only optional (or some optional) arguments. Some have a simplified nullary signature, such as `describeVpcs` (which might take an associated request object for a more detailed query). Some do not, even though their associated request object has only optional attributes, such as `describeEgressOnlyInternetGateways`. Some request classes specify that an attribute is optional, but they do so informally in the docstring. Either way, `describeEgressOnlyInternetGateways` doesn't.

An application can statically detect that there's a nullary version of a method, making it easy to deduce that it has only optional arguments. As far as I can tell, an application can not statically detect that `describeEgressOnlyInternetGateways` can be called with no arguments (i.e. it has no required arguments). In the case of the latter, a user ostensibly (I mean, I could be missing how) can't determine if its attribute is optional using the Javadoc either.

This information is available in the JSON API descriptions. As far as I can tell from the code gen macros, this isn't exposed in the API via e.g. an annotation. I think it should be: it would make it easier to use the Javadoc to write applications using the SDK manually, and it would also be available for programs to inspect.

Right now, I'm instead finding JAR files and extracting the JSON resource and parsing it. (This is necessary because I don't want to hardcode the version names, which are in the resource's name.) It'd be much more convenient for future users if they didn't have to do that; listing all available resources can be annoying.

**Comment by kiiadi** [comment]
@lvh great points all - we're constantly trying to improve our documentation to make the APIs easier to use. Currently the SDK does minimal validation of request properties and defers that logic to the service-side. This gives us maximum flexibility in terms of allowing the service to evolve without (necessarily) forcing customers to upgrade their SDK versions.

Whilst we know when we vend the SDK which service parameters are optional and which are required, this may change over time. A service for example, may relax a constraint making a previously required parameter optional. This could potentially change the signature of some methods or validation introducing non-backwards compatible changes - thus making it difficult for customers to upgrade their SDK versions - and so we defer that validation to the service as a general rule.

The service API documentation is the canonical source of what is and isn't required, but you're right at present it's difficult to get from the Javadoc to the service docs. The timing of this question is great - we're actually looking to add a link from Javadoc to the service documentation in our next SDK release (most likely today). Hopefully that'll allow you to more easily get from the SDK to the actual service documentation for a given operation.

**Comment by lvh** [comment]
@kiiadi Thanks for your reply! To clarify: I'm not suggesting the optional/required validation prevent you from even attempting a request; I am merely suggesting that that information be represented in the API.

Linking from the Javadoc would be an improvement, but only for humans on a documentation-reading-spree. My reasoning for suggesting an annotation is that it would help both humans in the Javadoc (depending on if that link looks like including a paragraph or a URL; this could remove another hop) as well as programs trying to parse the same information (e.g. IDEs).

While I'm not suggesting the gating validation, I'm not sure I understand the example of a relaxed requirement resulting in upgrading difficulties. An _old_ SDK might not let you make the new request (if it gates on validation), but a new SDK is presumably fine with you calling less `withQuantumTransmogrifierAttribute` methods on one of its request classes, and hence not setting a previously-required attribute, right? It seems like you'd mainly get backwards compatibility issues when going from optional to required.

**Comment by kiiadi** [comment]
You're right about having the main problem when going from optional to required - which we don't really allow services to do. I like the idea of an annotation - however the piece that worries me is that, the SDK is *not* the canonical source of what is and isn't required as I said - this belongs with the service team.

The other issue is that not all of our service-models accurately reflect how a service behaves (ie: what is and is not required). Obviously they should, and we do strive to get them as accurate as possible but we've seen cases in the past where things like the required attributes in the service models are not consistently populated (it's really more of an annotation in the model than a hard requirement). Hence you may have many attributes marked as Optional in the code/javadoc but that are in-fact required by the service. At the end of the day the service has the final say.

What's your specific use-case?

**Comment by lvh** [comment]
I'm trying to snapshot AWS service states so I can diff them and query the changes over time.

Since the code is generated from the model files, I'm not sure I understand why this changes what the canonical source of truth is. The SDK clearly already has opinions about what operations exist, what kind of arguments they take, et cetera. That doesn't make it the canonical source, but there are still all sorts of things e.g. the Java compiler can tell you about your call being wrong (wrong attribute name, for example). Whether an attribute is required doesn't seem categorically different. The operations that exist are also evidently the responsibility of the service team.

Similarly, having the canonical document be incorrect about which attributes are required or not doesn't seem particularly fundamentally different from the canonical document being incorrect about anything else -- it's a bug that just needs to be fixed, right? If you're just saying that they are not categorically different, but simply quantitatively much more likely to be wrong, that's something else entirely... But I for one will be happy to submit patches for all of the wrong ones I find ;)

(Relatedly: it seems like most SDKs have a vendored checkout of the same JSON source data. Is the canonical repo for those available? I haven't seen it; presumably that's where you'd want those patches to go.)

**Comment by kiiadi** [comment]
Required vs optional is an annotation that services may or may-not add to their model definitions. SDKs to these services can be built without this information - hence, as with anything that's not *required* it has the tendency to rot (AFAIK no SDK has validation against these attributes).

Operation names and parameters on the other hand are required. Calls to the service would fail if these weren't correct. As you put it, they are a different 'category' of information describing the API. Don't get me wrong, I'm not saying it's not important and service models should absolutely reflect the true state of the API - improving the accuracy of these models is something we're driving for.

In terms of the models they're mastered that's internal-only unfortunately; each service maintains it's on model. The JSON representation that many SDK's (including the Java SDK) vend with their releases (eg: `https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-codebuild/src/main/resources/models/codebuild-2016-10-06-model.json`) is transformed from the internal, proprietary format.

Would these JSON models give you what you need? They'd still suffer from the potential inaccuracy around required/optional - but they should give you a good sense of how the API is changing over time.

**Comment by lvh** [comment]
I'm reminded of the following quote:

> On two occasions I have been asked, "Pray, Mr. Babbage, if you put into the machine wrong figures, will the right answers come out?" ... I am not able rightly to apprehend the kind of confusion of ideas that could provoke such a question.
> — Charles Babbage, Passages from the Life of a Philosopher

;-)

I've been doing the "go grab the JSON" thing, but that is way more annoying to do than to just read the annotations (especially since I already have the reflected method objects anyway). I'm using Clojure and [amazonica][amazonica]; I've submitted a patch that exposes the underlying client & methods that is in the most current release. Then, it's still:

```clojure
(def jsons
  "Given a set of amazonica fn vars, find the JSON model resources in their jars."
  (comp
   (map (fn [fn-var]
          (-> fn-var meta :amazonica/client
              .getProtectionDomain .getCodeSource .getLocation)))
   (filter cp/jar-file?)
   (distinct)
   (mapcat (fn [loc] (-> loc io/file JarFile. cp/filenames-in-jar)))
   (filter (partial re-matches #"models/(\w+)-\d{4}-\d{2}-\d{2}-model.json"))))
```

(`cp` is `clojure.java.classpath`)

Then I have the JSONs, but I still don't have the required vs optional data. I can't just open the resource by name because the filenames have dates in them, so they're unpredictable (and enumerating the possible year, month, date values seems way grosser than what I'm doing above).

TL;DR, it's doable, but, ech. Having the annotation seems preferable for people looking at the Javadoc via the rendered HTML or via an IDE, even if it's sometimes wrong. As long as it's not wrong 80% of the time or something. I guess there's also a chicken-or-the-egg problem here: few people are interested in fixing the annotations because few people use them; and few people use them anyway because they're likely to be wrong :D

Again; I'm super happy to submit bugs about this, and I'll probably run into literally all of them for all idempotent/nilpotent operations :) Is there a useful way for me to contribute fixes to that internal thing? I think I know at least one person on an AWS SDK team personally; I think they start on Monday or something :D

[amazonica]: https://github.com/mcohen01/amazonica

**Comment by fernomac** [comment]
The data in the model is "correct" in that if an element is marked as required, then a request without it will definitely fail; this is enforced by the server-side deserializer (or at least was when I worked there). It's not necessarily complete though: some services include hand-written checks that a particular parameter is not null rather than mark the parameter required in the model. Sometimes this is an oversight, sometimes it's done intentionally to return a nicer error code/message than you'd get from the deserializer.

What really makes this tricky though is that the model has no way of representing more complex constraints like "at least one of X or Y is required" (they both have to be "optional" according to the model). Even assuming everything was annotated "correctly," there's no guarantee you can call an operation with no 'required' parameters without passing any parameters.

Is the backing source for the simplified method forms in the intermediate model public now? That's not complete as of now, but seems like the right place to crowdsource (possibly empty) sets of parameters that you might want to use for a given operation.

**Comment by lvh** [comment]
FWIW, I've been working on this more and hit a few snags. For example, even though the Java SDK consistently uses `FooBarResult` object naming, the shape in the original model will sometimes (not always) be called `FooBarResponse`. For example, the shape of the EC2 `DescribeVpcsResult` class is indeed called `DescribeVpcsResult`, but the shape of the Shield `DescribeAttackResult` class is called `DescribeAttackResponse`.

For future readers: this subtlety is why you probably want the `intermediate` model. I think. I'll keep you posted.

**Comment by shorea** [comment]
> For future readers: this subtlety is why you probably want the intermediate model. I think. I'll keep you posted.

The intermediate model is not something we make strong commitments on maintaining backwards compatibility. The schema of it may change in the future so tread carefully if using it. 

Is there a standard Java annotation to indicate something is required? Keep in mind we still support Java6.

**Comment by kiiadi** [comment]
It'd be interested if there was a standard "required" annotation but depending on its defined use I'd be a bit hesitant to use it given we have our own special definition which is really actually "probably required".

**Comment by lvh** [comment]
@shorea Is there a reasonable way to extract the same information from the model JSON? Say, given the example of the Result vs Response shape name inconsistency I mentioned.

**Comment by shorea** [comment]
If we do the annotation approach would you need the model?

For request response naming we go with OperationName + Request and OperationName + Result. So if the operation is DescribeInstances you'll have DescribeInstancesRequest and DescribeInstancesResult. There are a few exceptions in a couple of services though so it's not 100% reliable.

**Comment by lvh** [comment]
No, I don't think I would need the intermediate form if the annotations were present. I am using the intermediate form because it is the only thing that has both the `required` annotation and the correct Java types. (Now that I have them, I'm using them for other stuff, but I'm sure I could extract that from the reflection API as well.)

Re: exceptions to the rule in a handful of services: yes; that's my point exactly. The Intermediate form has the actual Java type names, so it doesn't get to lie. The model doesn't, so I need to guess.

**Comment by shorea** [comment]
I like the idea of annotations if just for documentation purposes so I'd prefer to go down that route.

If you want to use the intermediate model for other things than that's completely fine. Just be defensive with that code as the schema may change in incompatible ways in the future.

**Comment by lvh** [comment]
Gotcha, thanks :D Right now I mostly just care about `[service "shapes" result-class-name "required"]`; i.e. using it 100% like `...-model.json` except with the result classes spelled correctly. Hopefully that won't be changing often.

Any suggestions on where to report bugs in the models? I could open up tickets here, since there doesn't seem to be any specific better place.

**Comment by shorea** [comment]
Reporting here sounds good. We have to forward to the service team to fix but it's easiest to track on Github.

**Comment by spfink** [comment]
I'm still not convinced of the benefits of adding annotations due to the point that @fernomac brought up above that still is an issue with our modeling:

> What really makes this tricky though is that the model has no way of representing more complex constraints like "at least one of X or Y is required" (they both have to be "optional" according to the model).

Any required annotations we could use would not really accurately show what is required because of this.

**Comment by shorea** [comment]
Optionally required params are a pattern we generally discourage so I'm not as worried about that. 

Having the annotations for the things we know are required seems useful. It's not a 100% guarantee but it's something.

**Comment by spfink** [comment]
We would need to investigate to see if there are any well known standards in this area then. Typically  if you are going to include annotations in your models to let users know what is required I would assume that I could fill out the required values and have my request succeed. We simply can't make that statement.

The statement we will have to make will be something like:

"If all values marked with @Required are filled out your request may succeed. Please refer to the API documentation to determine if additional values are required for your parameter set"

vs a statement that would be much better but we can't make:

"If all values marked with @Required are filled out your request will succeed"

**Comment by lvh** [comment]
As a user who is now unusually familiar with almost all of the list/describe operations in all of AWS,  I agree. When this falls over, it's most commonly a categorical problem (e.g. S3, S3 transfer). It doesn't have to be accurate in 100% of cases to save a lot of time for the median user who no longer manually needs to go refer to the API documentations (when consuming the Javadoc) for every operation, and instead only for the few operations for which the required list is incorrect (and only if they incidentally use those).

By the way, is the inclusion of these model files in the JARs intentional? I'm not complaining -- it certainly saved my backside here -- but it seems like not a lot of users care, and those files can get pretty big. For example, the EC2 intermediate JSON alone is 17 MB. Fortunately, they compress extraordinarily well. Also, each of them is duplicated in the repository twice (where they are not compressed), which seems a little weird at first :)

**Comment by shorea** [comment]
There is no way we could make that statement even if we could rely on required. A call may fail because you don't have the appropriate resources, or the values of the required params are present but invalid. I think having the annotation gives the user a hint that they need to supply it but does not guarantee that's all they need to make it work.

**Comment by shorea** [comment]
> By the way, is the inclusion of these model files in the JARs intentional?

Which version of the SDK are you using? I thought we fixed that a few releases back. There is a separate module which contains all the model files now.

**Comment by lvh** [comment]
1.11.63, according to `lein deps :tree`:

```
   [com.amazonaws/aws-java-sdk "1.11.63" :exclusions [[joda-time]]]
     [com.amazonaws/aws-java-sdk-acm "1.11.63"]
     [com.amazonaws/aws-java-sdk-api-gateway "1.11.63"]
     [com.amazonaws/aws-java-sdk-applicationautoscaling "1.11.63"]
     [com.amazonaws/aws-java-sdk-appstream "1.11.63"]
     [com.amazonaws/aws-java-sdk-autoscaling "1.11.63"]
     [com.amazonaws/aws-java-sdk-budgets "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudformation "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudfront "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudhsm "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudsearch "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudtrail "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudwatch "1.11.63"]
     [com.amazonaws/aws-java-sdk-cloudwatchmetrics "1.11.63"]
     [com.amazonaws/aws-java-sdk-codebuild "1.11.63"]
     [com.amazonaws/aws-java-sdk-codecommit "1.11.63"]
     [com.amazonaws/aws-java-sdk-codedeploy "1.11.63"]
     [com.amazonaws/aws-java-sdk-codepipeline "1.11.63"]
     [com.amazonaws/aws-java-sdk-cognitoidentity "1.11.63"]
     [com.amazonaws/aws-java-sdk-cognitoidp "1.11.63"]
     [com.amazonaws/aws-java-sdk-cognitosync "1.11.63"]
     [com.amazonaws/aws-java-sdk-config "1.11.63"]
     [com.amazonaws/aws-java-sdk-core "1.11.63"]
     [com.amazonaws/aws-java-sdk-datapipeline "1.11.63"]
     [com.amazonaws/aws-java-sdk-devicefarm "1.11.63"]
     [com.amazonaws/aws-java-sdk-directconnect "1.11.63"]
     [com.amazonaws/aws-java-sdk-directory "1.11.63"]
     [com.amazonaws/aws-java-sdk-discovery "1.11.63"]
     [com.amazonaws/aws-java-sdk-dms "1.11.63"]
     [com.amazonaws/aws-java-sdk-dynamodb "1.11.63"]
     [com.amazonaws/aws-java-sdk-ec2 "1.11.63"]
     [com.amazonaws/aws-java-sdk-ecr "1.11.63"]
     [com.amazonaws/aws-java-sdk-ecs "1.11.63"]
     [com.amazonaws/aws-java-sdk-efs "1.11.63"]
     [com.amazonaws/aws-java-sdk-elasticache "1.11.63"]
     [com.amazonaws/aws-java-sdk-elasticbeanstalk "1.11.63"]
     [com.amazonaws/aws-java-sdk-elasticloadbalancing "1.11.63"]
     [com.amazonaws/aws-java-sdk-elasticloadbalancingv2 "1.11.63"]
     [com.amazonaws/aws-java-sdk-elasticsearch "1.11.63"]
     [com.amazonaws/aws-java-sdk-elastictranscoder "1.11.63"]
     [com.amazonaws/aws-java-sdk-emr "1.11.63"]
     [com.amazonaws/aws-java-sdk-events "1.11.63"]
     [com.amazonaws/aws-java-sdk-gamelift "1.11.63"]
     [com.amazonaws/aws-java-sdk-glacier "1.11.63"]
     [com.amazonaws/aws-java-sdk-health "1.11.63"]
     [com.amazonaws/aws-java-sdk-iam "1.11.63"]
     [com.amazonaws/aws-java-sdk-importexport "1.11.63"]
     [com.amazonaws/aws-java-sdk-inspector "1.11.63"]
     [com.amazonaws/aws-java-sdk-iot "1.11.63"]
     [com.amazonaws/aws-java-sdk-kinesis "1.11.63"]
     [com.amazonaws/aws-java-sdk-kms "1.11.63"]
     [com.amazonaws/aws-java-sdk-lambda "1.11.63"]
     [com.amazonaws/aws-java-sdk-lightsail "1.11.63"]
     [com.amazonaws/aws-java-sdk-logs "1.11.63"]
     [com.amazonaws/aws-java-sdk-machinelearning "1.11.63"]
     [com.amazonaws/aws-java-sdk-marketplacecommerceanalytics "1.11.63"]
     [com.amazonaws/aws-java-sdk-marketplacemeteringservice "1.11.63"]
     [com.amazonaws/aws-java-sdk-models "1.11.63"]
     [com.amazonaws/aws-java-sdk-opsworks "1.11.63"]
     [com.amazonaws/aws-java-sdk-opsworkscm "1.11.63"]
     [com.amazonaws/aws-java-sdk-pinpoint "1.11.63"]
     [com.amazonaws/aws-java-sdk-polly "1.11.63"]
     [com.amazonaws/aws-java-sdk-rds "1.11.63"]
     [com.amazonaws/aws-java-sdk-redshift "1.11.63"]
     [com.amazonaws/aws-java-sdk-rekognition "1.11.63"]
     [com.amazonaws/aws-java-sdk-route53 "1.11.63"]
     [com.amazonaws/aws-java-sdk-s3 "1.11.63"]
     [com.amazonaws/aws-java-sdk-servermigration "1.11.63"]
     [com.amazonaws/aws-java-sdk-servicecatalog "1.11.63"]
     [com.amazonaws/aws-java-sdk-ses "1.11.63"]
     [com.amazonaws/aws-java-sdk-shield "1.11.63"]
     [com.amazonaws/aws-java-sdk-simpledb "1.11.63"]
     [com.amazonaws/aws-java-sdk-simpleworkflow "1.11.63"]
     [com.amazonaws/aws-java-sdk-snowball "1.11.63"]
     [com.amazonaws/aws-java-sdk-sns "1.11.63"]
     [com.amazonaws/aws-java-sdk-sqs "1.11.63"]
     [com.amazonaws/aws-java-sdk-ssm "1.11.63"]
     [com.amazonaws/aws-java-sdk-stepfunctions "1.11.63"]
     [com.amazonaws/aws-java-sdk-storagegateway "1.11.63"]
     [com.amazonaws/aws-java-sdk-sts "1.11.63"]
     [com.amazonaws/aws-java-sdk-support "1.11.63"]
     [com.amazonaws/aws-java-sdk-swf-libraries "1.11.22"]
     [com.amazonaws/aws-java-sdk-waf "1.11.63"]
     [com.amazonaws/aws-java-sdk-workspaces "1.11.63"]
     [com.amazonaws/aws-java-sdk-xray "1.11.63"]
```

As you can see, I have the models jar, too. However, I downloaded the most recent EC2 one off of Maven Central:

```
lvh@plotka ~/D/java-sdk> ls
aws-java-sdk-ec2-1.11.75.jar
lvh@plotka ~/D/java-sdk> jar -tf *jar | grep json
models/ec2-2016-11-15-model.json
models/ec2-2016-11-15-intermediate.json
```

Taking the JSON out and repacking shaves 1MB off:

```
-rw-r--r--@ 1 lvh  staff   3.8M Jan  3 12:01 aws-java-sdk-ec2-1.11.75.jar
-rw-r--r--  1 lvh  staff   2.8M Jan  3 12:12 new.jar
```

**Comment by lvh** [comment]
I happen to know I'm getting them from the individual service JARs, because the way I'm finding the models involves finding the JAR that I got the client class from, and then associating whatever I find in that jar with the client. That works fine for now, but I should probably change that anyway since it would also fail e.g. when producing an uberjar :)

**Comment by shorea** [comment]
Ah I see. That's a regression then. They should only be present in aws-java-sdk-models.

**Comment by lvh** [comment]
nooo please don't take my precious jsons until I've changed my coooode

(jk)

**Comment by debora-ito** [comment]
Hi @lvh, the SDK team has reviewed the feature request list for V1, and since they're concentrating efforts on V2 new features they decided to not implement this one in V1. One of the reasons is it would not be possible to maintain backwards compatibility. I'll go ahead and close this.

Please feel free to reach out if you have further questions.

**Comment by lvh** [comment]
I'm not sure I follow. Isn't this a property of the model files, not of the specific SDK version? (Also, doesn't the problem exist identically in "V2"?)

**Comment by zoewangg** [comment]
Sorry for being unclear. Yes, it would be a property on the model files, however, the model files are owned and maintained by the service teams and they might change the constraint on certain fields and that could introduce breaking changes.

Yes, we have the same feature request on v2 but it's not on our roadmap currently. https://github.com/aws/aws-sdk-java-v2/issues/47

**Comment by lvh** [comment]
Sorry, that first paragraph makes sense but I don't understand how it applies to this ticket. Sure, teams can make breaking changes to model files -- how does that affect the expressiveness of the model files to describe which attributes are optional? Also, the ticket you linked is about returning nulls, which AFAICT isn't responsive to the question asked by this ticket.
